gel/protocol/protocol.pyx

```diff
@@ -553,6 +645,137 @@ cdef class SansIOProtocol:
+                    self.parse_describe_state_message()
+
+                elif mtype == DATA_MSG:
+                    if exc is None:
+                        try:
+                            self.parse_data_messages(ctx.out_dc, result)
+                        except Exception as ex:
+                            # An error during data decoding.  We need to
+                            # handle this as gracefully as possible:
+                            # * save the exception to raise it once SYNC is
+                            #   received;
+                            # * ignore all 'D' messages for this query.
+                            exc = errors.ClientError(
+                                'unable to decode data to Python objects')
+                            exc.__cause__ = ex
+                            # Take care of a partially consumed 'D' message
+                            # and the ones yet unparsed.
+                            while self.buffer.take_message_type(DATA_MSG):
+                                self.buffer.discard_message()
+                    else:
+                        self.buffer.discard_message()
+
+                elif mtype == COMMAND_COMPLETE_MSG:
+                    self.parse_command_complete_message()
+                    results.append(result)
+                    result = []
+                    if ctx is ctxs[-1]:
+                        ctx = None
+                    else:
+                        ctx = next(it)
+
+                elif mtype == ERROR_RESPONSE_MSG:
+                    exc = self.parse_error_message()
+                    exc._query = ctx.query
+                    if exc.get_code() == parameter_type_mismatch_code:
+                        if not isinstance(ctx.in_dc, NullCodec):
+                            buf = WriteBuffer.new()
+                            try:
+                                self.encode_args(
+                                    ctx.in_dc, buf, ctx.args, ctx.kwargs
+                                )
+                            except errors.QueryArgumentError as ex:
+                                exc = ex
+                            else:
+                                # Now we know for sure that the new codec can
+                                # encode the same arguments, let's hint the
+                                # caller to retry as a special case.
+                                exc.tags = exc.tags.union(
+                                    {errors.SHOULD_RETRY}
+                                )
+
+                                # If we may retry this batch, invalidate codec
+                                # cache for all queries except the current one
+                                # because we just updated its input codec
+                                for c in ctxs:
+                                    if c is not ctx:
+                                        c.invalidate_cache()
```

Comment:
We need to exit the loop or re-raise? Not sure I see what happens next after invalidation here